src/util/obbt.jl

```diff
@@ -0,0 +1,406 @@
+export run_obbt_opf
+
+"optimality-based bound tightening for Optimal Power Flow Relaxations"
+
+
+function check_variables(pm::GenericPowerModel) 
+    try 
+        vm = var(pm, :vm)
+    catch err 
+        (isa(error, KeyError)) && (error(LOGGER, "OBBT is not supported for models without explicit voltage magnitude variables"))
+    end 
+
+    try 
+        td = var(pm, :td)
+    catch err 
+        (isa(error, KeyError)) && (error(LOGGER, "OBBT is not supported for models without explicit voltage angle difference variables"))
+    end 
+end 
+
+function check_obbt_options(ub::Float64, rel_gap::Float64, ub_constraint::Bool)
+    if ub_constraint && isinf(ub)
+        error(LOGGER, "the option upper_bound_constraint cannot be set to true without specifying an upper bound")
+    end 
+
+    if !isinf(rel_gap) && isinf(ub)
+        error(LOGGER, "rel_gap_tol is specified without providing an upper bound")
+    end
+end 
+
+function constraint_obj_bound_polynomial(pm::GenericPowerModel, bound)
+    PowerModels.check_polynomial_cost_models(pm)
+
+    from_idx = Dict(arc[1] => arc for arc in ref(pm, :arcs_from_dc))
+
+    @constraint(pm.model, 
+            sum(   
+                gen["cost"][1]*var(pm, :pg, i)^2 +
+                gen["cost"][2]*var(pm, :pg, i) +
+                gen["cost"][3]
+            for (i,gen) in ref(pm, :gen)) +
+            sum(   
+                dcline["cost"][1]*var(pm,:p_dc, from_idx[i])^2 +
+                dcline["cost"][2]*var(pm, :p_dc, from_idx[i])  +
+                dcline["cost"][3] 
+            for (i,dcline) in ref(pm, :dcline))
+            <= bound 
+    )
+end 
+
+function constraint_obj_bound(pm::GenericPowerModel, bound)
+    model = PowerModels.check_cost_models(pm)
+    if model != 2 
+        error("Only cost models of type 2 is supported at this time, given cost model type $(model)")
+    end 
+    constraint_obj_bound_polynomial(pm, bound)
```

Comment:
No need to break this out as a separate function.

Reply:
converted to a single function